subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/transaction/CompileTransaction.java

```diff
@@ -0,0 +1,302 @@
+        ensureEmptyDirectoriesBeforeExecution(stagedOutputs);
+        StashResult stashResult = stashFilesThatShouldBeDeleted();
+        try {
+            setupSpecOutputs(stagedOutputs);
+            T result = function.apply(stashResult.mapToWorkResult());
+            deletePotentiallyEmptyDirectories(stashResult);
+            moveCompileOutputToOriginalFolders(stagedOutputs);
+            return result;
+        } catch (Exception t) {
+            rollbackStash(stashResult.stashedFiles);
+            throw t;
+        } finally {
+            restoreSpecOutputs(stagedOutputs);
+        }
+    }
+
+    private void ensureEmptyDirectoriesBeforeExecution(List<StagedOutput> stagedOutputs) {
+        try {
+            tempDir.mkdirs();
+
+            // Create or clean stash and stage directories
+            Set<File> ensureEmptyDirectories = new HashSet<>();
+            deleter.ensureEmptyDirectory(stashDirectory);
+            ensureEmptyDirectories.add(stashDirectory);
+            for (StagedOutput output : stagedOutputs) {
+                ensureEmptyKeepingFolderStructure(output);
+                ensureEmptyDirectories.add(output.stagingDirectory);
+            }
+
+            // Delete any other file or directory
+            try (Stream<Path> dirStream = Files.list(tempDir.toPath())) {
+                dirStream.map(Path::toFile)
+                    .filter(file -> !ensureEmptyDirectories.contains(file))
+                    .forEach(this::deleteRecursively);
+            }
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private void ensureEmptyKeepingFolderStructure(StagedOutput output) throws IOException {
+        Path currentDir = output.stagingDirectory.toPath();
+        if (!Files.exists(currentDir)) {
+            Files.createDirectory(currentDir);
+            return;
+        }
+        try (Stream<Path> dirStream = Files.walk(currentDir)) {
+            // Delete all files and delete all directories
+            // that don't exist in sourceDirectory
+            dirStream
+                // Order files in a direction that we can avoid recursive deletion
+                .sorted(Comparator.reverseOrder())
+                .filter(path -> !Files.isDirectory(path) || !isDirectoryAlsoInOtherRoot(path, currentDir, output.sourceDirectory))
+                .forEach(path -> path.toFile().delete());
+        }
+    }
+
+    private boolean isDirectoryAlsoInOtherRoot(Path directory, Path root, File otherRoot) {
+        Path relativePath = root.relativize(directory);
+        File fileInOtherRoot = new File(otherRoot, relativePath.toString());
+        return fileInOtherRoot.isDirectory();
+    }
+
+    private void deleteRecursively(File file) {
+        try {
+            deleter.deleteRecursively(file);
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private List<StagedOutput> collectOutputsToStage() {
```

Comment:
So I think here the main concern are annotation processors. If annotation processor is not incremental, then there is no issue, we always do full recompilation.

If it is incremental I think there are two cases:
1. If it writes to some output folders that are not registered, then it's its responsibility to handle this. (I hope there is no such processor)
2. If it generates something to some registered output folder (that is not compile output/generated sources) then output will be dirty and we will do full recompilation in case of a failure as before.

To answer
> Do we need some way that a user can add custom output locations here? 

I would try to avoid, and I hope annotation processors writes to standard generated sources dir or standard compile output dir. But if there are some annotation processors that benefit from that, we might want to add it. If you know for any such case let me know.

Reply:
I'd also prefer not to add custom output folders. Let's leave this as is. I wonder if we could get some early feedback from builds which use annotation processors.